src/celeritas/optical/gen/DirectGeneratorAction.cc

```diff
@@ -0,0 +1,237 @@
+    GeneratorRegistry& gen = *params.gen_reg();
+    auto result = std::make_shared<DirectGeneratorAction>(
+        actions.next_id(), aux.next_id(), gen.next_id());
+    actions.insert(result);
+    aux.insert(result);
+    gen.insert(result);
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with action and data IDs.
+ */
+DirectGeneratorAction::DirectGeneratorAction(ActionId id,
+                                             AuxId aux_id,
+                                             GeneratorId gen_id)
+    : GeneratorBase(id,
+                    aux_id,
+                    gen_id,
+                    "generate-direct",
+                    "directly generate optical photon primaries")
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Insert user-provided host initializer data.
+ */
+void DirectGeneratorAction::insert(CoreStateBase& state,
+                                   SpanConstData data) const
+{
+    if (auto* s = dynamic_cast<CoreStateHost*>(&state))
+    {
+        return this->insert_impl(*s, data);
+    }
+    else if (auto* s = dynamic_cast<CoreStateDevice*>(&state))
+    {
+        return this->insert_impl(*s, data);
+    }
+    CELER_ASSERT_UNREACHABLE();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Execute the action with host data.
+ */
+void DirectGeneratorAction::step(CoreParams const& params,
+                                 CoreStateHost& state) const
+{
+    this->step_impl(params, state);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Execute the action with device data.
+ */
+void DirectGeneratorAction::step(CoreParams const& params,
+                                 CoreStateDevice& state) const
+{
+    this->step_impl(params, state);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Add initializers to the aux state.
+ */
+template<MemSpace M>
+void DirectGeneratorAction::insert_impl(CoreState<M>& state,
+                                        SpanConstData data) const
+{
+    CELER_EXPECT(state.aux());
+
+    auto& aux_state
+        = get<DirectGeneratorState<M>>(*state.aux(), this->aux_id());
+
+    if (aux_state.counters.buffer_size != 0)
+    {
+        CELER_NOT_IMPLEMENTED(
+            "multiple consecutive direct photon insertions not supported");
+    }
+
+    if (aux_state.store.size() < data.size())
+    {
+        // Reallocate with enough capacity
+        aux_state.store = CollectionStateStore<DirectGeneratorStateData, M>{
+            state.stream_id(), static_cast<size_type>(data.size())};
+    }
+
+    // Update counters and copy distributions to aux state storage
+    aux_state.counters.buffer_size = data.size();
+    aux_state.counters.num_pending = data.size();
+    state.counters().num_pending = data.size();
```

Comment:
```suggestion
    state.counters().num_pending += data.size();
```
Since this is number of pending photons over all the generators. (My bad since I made this mistake in the `PrimaryGeneratorAction`: I'll fix that.)